cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringAgentService.java

```diff
@@ -188,11 +188,23 @@ protected long runTask() {
             handleResponse(resp, peer);
             break;
           case HttpURLConnection.HTTP_NOT_FOUND:
-            handleNotFound(peer);
-            break;
+            if (resp.getResponseBodyAsString(StandardCharsets.UTF_8)
+              .equals(String.format("Peer %s not found", instanceName))) {
```

Comment:
we don't want to be using message strings to control logic. It makes it very easy for somebody to change a string and unknowingly break the other side.

Instead of direct url and string manipulation here, there should be a TetheringClient layer that handles serializing/deserializing requests and responses into actual java classes and throws appropriate exceptions.

Reply:
fixed